homeassistant/components/calendar/__init__.py

```diff
@@ -0,0 +1,286 @@
+        start = self._cal_data.get('start', None)
+        end = self._cal_data.get('end', None)
+        start = start.isoformat() if start is not None else None
+        end = end.isoformat() if end is not None else None
+
+        return {
+            'message': self._cal_data.get('message', ''),
+            'all_day': self._cal_data.get('all_day', False),
+            'offset_reached': self.offset_reached,
+            'start_time': start,
+            'end_time': end,
+            'location': self._cal_data.get('location', None),
+            'description': self._cal_data.get('description', None),
+        }
+
+    @property
+    def track(self):
+        """Are we tracking events on this search sensor."""
+        return self._track
+
+    @property
+    def is_all_day(self):
+        """If the next event is all day then True."""
+        return self._cal_data.get('all_day', False)
+
+    @property
+    def search(self):
+        """Return the search string."""
+        return self._search
+
+    @property
+    def name(self):
+        """Return the device name."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Return if the event in process."""
+        return self._state
+
+    @property
+    def state(self):
+        """Return the state of the calendar event."""
+        return STATE_ON if self.is_on else STATE_OFF
+
+    def cleanup(self):
+        """Cleanup any start/end listeners that were setup."""
+        from homeassistant.const import EVENT_TIME_CHANGED
+        self._cal_data = {
+            'all_day': False,
+            'offset_time': 0,
+            'message': '',
+            'start': None,
+            'end': None,
+            'location': None,
+            'description': None
+        }
+
+        if self._start_listener:
+            self.hass.bus.remove_listener(EVENT_TIME_CHANGED,
+                                          self._start_listener)
+            self._start_listener = None
+        if self._end_listener:
+            self.hass.bus.remove_listener(EVENT_TIME_CHANGED,
+                                          self._end_listener)
+            self._end_listener = None
+
+    # pylint: disable=no-self-use
+    def get_next_event(self):
+        """Raise an error so that subclasses implement this."""
+        message = """Should return:
+        event_allday = {
+            'start': {'date': '2016-01-30'},
+            'end': {'date': '2016-01-31'},
+            'summary': 'Event Title',
+            'location': 'Location',
+            'description': 'Notes of event',
+        }
+        OR
+        event = {
+            'start': {'dateTime': '2016-01-30T01:00:00-00:00'},
+            'end': {'dateTime': '2016-01-30T02:00:00-00:00'},
+            'summary': 'Event Title',
+            'location': 'Location',
+            'description': 'Notes of event',
+        }
+        """
+        raise NotImplementedError(message)
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
```

Comment:
We should never throttle at the entity level. It is up to the platforms to decide how much they require throttling. For example the demo platform will never need a throttle.